### PR TITLE
Kyle/gfc 791/cache tax period

### PR DESCRIPTION
**Pull request by kylemchownhmrc**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/sharedmodel/TaxPeriods.scala [hunk 1]
```diff
@@ -49,3 +50,41 @@ case class TaxResponse(id: HmrcTaxPeriod, obligation: Obligation)
...
   implicit val format: OFormat[TaxResponse] = Json.format[TaxResponse]
 }
+
+case class TaxPeriodIdentifier(idType: IdType, idNumber: IdNumber, regimeType: RegimeType)
+
+object TaxPeriodIdentifier {
+  implicit val format: OFormat[TaxPeriodIdentifier] = Json.format[TaxPeriodIdentifier]
+}
+
+case class AllInfo(
```
Comment (lutnos) [inline app/uk/gov/hmrc/gform/sharedmodel/TaxPeriods.scala]: This name needs to be better.  Perhaps something along the lines of TaxPeriodMapping

File: app/uk/gov/hmrc/gform/sharedmodel/form/Form.scala [hunk 2]
```diff
@@ -41,7 +41,8 @@ case class Form(
   formData: FormData,
   status: FormStatus,
   visitsIndex: VisitIndex,
-  envelopeExpiryDate: Option[EnvelopeExpiryDate]
+  envelopeExpiryDate: Option[EnvelopeExpiryDate],
+  obligations: Option[ListAllInfo]
```
Comment (lutnos) [inline app/uk/gov/hmrc/gform/sharedmodel/form/Form.scala]: Is there a meaningful difference between None and Some(List.empty) ?  If not, then it should either not be an Option, or should be an Option of a NonEmptyList

Comment (lutnos) [reply app/uk/gov/hmrc/gform/sharedmodel/form/Form.scala]: In fact, this ought not be either a List[], or an Option[NonEmptyList], it should be

obligations: NonEmptyList[TaxPeriodInformation]

File: app/uk/gov/hmrc/gform/sharedmodel/TaxPeriods.scala [hunk 3]
```diff
@@ -49,3 +50,28 @@ case class TaxResponse(id: HmrcTaxPeriod, obligation: Obligation)
...
+  periodKey: String)
+
+object TaxPeriodInformation {
+  implicit val format: OFormat[TaxPeriodInformation] = derived.oformat
+}
+
+sealed trait Obligations
+final case object NotChecked extends Obligations
+final case object NoObligations extends Obligations
+final case class RetrievedObligations(listOfObligations: List[TaxPeriodInformation]) extends Obligations
```
Comment (lutnos) [inline app/uk/gov/hmrc/gform/sharedmodel/TaxPeriods.scala]: That needs to be a NonEmptyList[TaxPeriodInformation]]

File: app/uk/gov/hmrc/gform/formtemplate/FormComponentMaker.scala [hunk 4]
```diff
@@ -322,7 +322,7 @@ class FormComponentMaker(json: JsValue) {
 
   private lazy val hmrcTaxPeriodOpt: Opt[HmrcTaxPeriod] = {
     val oHmrcTaxPeriod: Opt[HmrcTaxPeriod] = (idType, idNumber, regimeType) match {
-      case (Some(a), Some(b), Some(c)) => HmrcTaxPeriod(IdType(a), IdNumber(b), RegimeType(c)).asRight
+      case (Some(a), Right(Some(b: TextExpression)), Some(c)) => HmrcTaxPeriod(IdType(a), b, RegimeType(c)).asRight
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/formtemplate/FormComponentMaker.scala]: What about all other cases? We are not handling them. I've never saw this from the compiler:

```
[warn]  [E3] app/uk/gov/hmrc/gform/formtemplate/FormComponentMaker.scala:324:46:
[warn]       Exhaustivity analysis reached max recursion depth, not all missing cases are reported.
[warn]       (Please try with scalac -Ypatmat-exhaust-depth 40 or -Ypatmat-exhaust-depth off.)
[warn]           val oHmrcTaxPeriod: Opt[HmrcTaxPeriod] = (idType, idNumber, regimeType) match {
```

**Comment by josef-vlach** [comment]
@lutnos @kylemchownhmrc How come this got merged? I don't think solution in [`6d8939d`](/hmrc/gform/commit/6d8939d9ea07dba2da987783c23df21694e1a91b) for my comment is correct, quite the contrary actually it make things even worse.

**Comment by lutnos** [comment]
@josef-vlach 

@kylemchownhmrc and I discussed recalculation here & we now believe we understand the problem and a proper solution for hmrcTaxPeriod .  We have it on a whiteboard and can share it with you tomorrow.  It will need a more that trial change.

For now, I have merged, although there may be a remaining issue with an empty NEL & passing an Expression to the backend is unsatisfactory.

I have done this because the current code will satisfy the AC for GFC-791, that an agent can enter a registration number and get an HmrcTaxPeriod.  This will satisfy our immediate requirement to have an automated Aggregate Levy return working.

As I say, we would like to discuss with you tomorrow

**Pull request merged**